File: src/test/java/station/CentralStationTest.java
package station;

import org.junit.jupiter.api.Test;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static station.DataStationTest.createData;
import static station.DataStationTest.createStation;
import static util.Util.matrixDiagonalMultiplication;

public class CentralStationTest {

    @Test
    public void testCalculateNPartyScalarProduct() {
        for (int n = 2; n < 8; n++) {
            for (int population = 2; population < 10; population++) {
                List<DataStation> datastations = new ArrayList<>();
                List<BigInteger[][]> datasets = new ArrayList<>();
                for (int i = 0; i < n; i++) {
                    BigInteger[][] data = createData(population);
                    datastations.add(new DataStation(String.valueOf(i), data));
                    datasets.add(data);
                }
                CentralStation central = new CentralStation();

                // calculate expected anwser:
                BigInteger expected = matrixDiagonalMultiplication(datasets, datasets.get(0).length);
                BigInteger result = central.calculateNPartyScalarProduct(datastations);

                assertEquals(expected, result);
            }
        }
    }

    @Test
    public void testCalculate3PartyScalarProduct() {
        final int POPUlATION = 5;
        final int N = 3;

        List<DataStation> datastations = new ArrayList<>();
        List<BigInteger[][]> datasets = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            BigInteger[][] data = createData(POPUlATION);
            datastations.add(new DataStation(String.valueOf(i), data));
            datasets.add(data);
        }
        CentralStation central = new CentralStation();

        // calculate expected anwser:
        BigInteger expected = matrixDiagonalMultiplication(datasets, datasets.get(0).length);
        BigInteger result = central.calculateNPartyScalarProduct(datastations);

        assertEquals(expected, result);
    }

    @Test
    public void testDetermineSubprotocols() {
        for (int population = 1; population < 10; population++) {
            List<DataStation> datastations = new ArrayList<>();
            for (int i = 0; i < population; i++) {
                datastations.add(createStation(String.valueOf(i), population));
            }
            CentralStation central = new CentralStation();
            SecretStation secretStation = new SecretStation();
            secretStation.shareSecret(datastations);

            List<List<DataStation>> subprotocols = central.determineSubprotocols(datastations, secretStation);
            assertEquals(subprotocols.size(), calculateExpectedCombinations(population).longValue());
        }
    }

    //calculates  n! / (x!(n−x)!) for every x >= 2 < n
    private BigInteger calculateExpectedCombinations(int n) {
        BigInteger nFactorial = factorial(n);
        BigInteger sum = BigInteger.ZERO;
        for (int x = 2; x < n; x++) {
            BigInteger xFactorial = factorial(x);
            BigInteger nxFactorial = factorial(n - x);
            sum = sum.add(nFactorial.divide(xFactorial.multiply(nxFactorial)));
        }
        return sum;
    }

    private BigInteger factorial(int n) {
        BigInteger nFactorial = BigInteger.ONE;
        for (int i = 1; i <= n; i++) {
            nFactorial = nFactorial.multiply(BigInteger.valueOf(i));
        }
        return nFactorial;
    }

}

File: python/test/station/CentralStation_test.py
import unittest

from station import CentralStation
from station import DataStation
from station import SecretStation
from test.station import DataStation_test
from util.Util import matrixDiagonalMultiplication


class CentralStationTest(unittest.TestCase):

    def test_CalculateNPartyScalarProduct(self):
        for n in range(2,8):
            for population in range(2,10):
                datastations = []
                datasets = []
                for i in range(0,n):
                    data = DataStation_test.createData(population)
                    datastations.append(DataStation.DataStation(str(i), data))
                    datasets.append(data)

                central = CentralStation()

                # calculate expected anwser:
                expected = matrixDiagonalMultiplication(datasets)
                result = central.calculateNPartyScalarProduct(datastations)

                self.assertEquals(expected, result)
                
    def test_Calculate3PartyScalarProduct(self):
        POPUlATION = 5
        N = 3

        datastations = []
        datasets = []
        for i in range(0,N):
            data = DataStation_test.createData(POPUlATION)
            datastations.append(DataStation.DataStation(str(i), data))
            datasets.append(data)
        
        central = CentralStation()

        # calculate expected anwser:
        expected = matrixDiagonalMultiplication(datasets)
        result = central.calculateNPartyScalarProduct(datastations)

        self.assertEquals(expected, result)

    def test_DetermineSubprotocols(self):
        for population in range(1, 10):
            datastations = []
            for i in range(0, population):
                datastations.append(DataStation_test.createStation(i, population))

            central = CentralStation()
            secretStation = SecretStation()
            secretStation.shareSecret(datastations)

            subprotocols = central.determineSubprotocols(datastations, secretStation)
            self.assertEquals(len(subprotocols), calculateExpectedCombinations(population))

def calculateExpectedCombinations(n:int ):
    nFactorial = factorial(n)
    sum = 0
    for x in range(2,n):
        xFactorial = factorial(x)
        nxFactorial = factorial(n - x)
        sum = sum + (nFactorial / (xFactorial *(nxFactorial)))

    return sum


def factorial(n: int):
    nFactorial = 1
    for i in range(1,n+1):
        nFactorial = nFactorial * i
    return nFactorial
